### PR TITLE
[mono][aot] Fix the inclusion of generic instances when doing profile…

### PR DESCRIPTION
**Pull request by vargaz**
…d AOT.

If a profile contained a method like List<AnInst>:.ctor () and the
'profile-only' option was used, the fully shared version of the method
needs to be added to the aot image.

### CODE REVIEW

File: src/mono/mono/mini/aot-compiler.c [hunk 1]
```diff
@@ -12909,8 +12909,15 @@ resolve_class (ClassProfileData *cdata)
 	}
 	if (cdata->inst) {
 		resolve_ginst (cdata->inst);
-		if (!cdata->inst->inst)
+		if (!cdata->inst->inst) {
+			/*
+			 * The instance might not be found if its arguments are in another assembly,
+			 * use the definition instead.
```
Comment (lambdageek) [inline src/mono/mono/mini/aot-compiler.c:12915]: How does this happen?   If an assembly is loaded at runtime and then an instance is created, for example?

File: src/mono/mono/mini/aot-compiler.c [hunk 2]
```diff
@@ -13021,9 +13030,11 @@ resolve_profile_data (MonoAotCompile *acfg, ProfileData *data, MonoAssembly* cur
 			char *sig_str = mono_signature_full_name (sig);
 			gboolean match = !strcmp (sig_str, mdata->signature);
 			g_free (sig_str);
-			if (!match)
-
-				continue;
+			if (!match) {
+				// The signature might not match for methods on gtds
+				if (!mono_class_is_gtd (klass))
```
Comment (lambdageek) [inline src/mono/mono/mini/aot-compiler.c:13035]: Is this going to be a problem with overloaded methods?  should we at least match the number of arguments?

**Review by lambdageek — APPROVED** [review]

File: src/mono/mono/mini/aot-compiler.c (hunk 1 above)
Comment (vargaz) [reply src/mono/mono/mini/aot-compiler.c:12915]: The profile might include methods like List<AClass>, and when corlib is AOT-ed, AClass-es assembly is not loaded.

File: src/mono/mono/mini/aot-compiler.c (hunk 2 above)
Comment (vargaz) [reply src/mono/mono/mini/aot-compiler.c:13035]: We do match above.

**Pull request merged**